File: Twt/Twt/BrowserViewController.m
//
//  BrowserViewController.m
//  Twt
//
//  Created by Nattapong Mos on 25/5/57.
//  Copyright (c) พ.ศ. 2557 Nattapong Mos. All rights reserved.
//

#import "BrowserViewController.h"
#import "NewTweetViewController.h"
#import "UIViewController+RequiredObject.h"
#import <SafariServices/SafariServices.h>

@interface BrowserViewController ()

@property (weak, nonatomic) IBOutlet UIWebView *webView;
@property (strong, nonatomic) IBOutlet UIBarButtonItem *activityItem;
@property (strong, nonatomic) IBOutlet UIToolbar *toolbar;
@property (strong, nonatomic) IBOutlet UIBarButtonItem *reloadItem;
@property (strong, nonatomic) IBOutlet UIBarButtonItem *stopItem;
@property (strong, nonatomic) IBOutlet UIBarButtonItem *backButton;
@property (strong, nonatomic) IBOutlet UIBarButtonItem *nextButton;
@property (strong, nonatomic) IBOutlet UIBarButtonItem *shareButton;
@property (strong, nonatomic) NSURL *url1;
@property (strong, nonatomic) NSURL *url2;
@property (strong, nonatomic) AppDelegate *appDelegate;
@property (strong, nonatomic) UIActivityIndicatorView *activityIndicator;

@end

@implementation BrowserViewController

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    if (self.backBarButtonItemTitle != nil) {
        UIBarButtonItem *backButton = [[UIBarButtonItem alloc] initWithTitle:self.backBarButtonItemTitle style:UIBarButtonItemStylePlain target:nil action:nil];
        self.navigationController.navigationBar.topItem.backBarButtonItem = backButton;
    }
    self.title = self.address.host;
    self.activityIndicator = [[UIActivityIndicatorView alloc] initWithFrame:CGRectMake(0, 0, 20, 20)];
    self.activityIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyleGray;
    [self.activityItem setCustomView:self.activityIndicator];
    [self.webView.scrollView setContentInset:UIEdgeInsetsMake(0, 0, 44, 0)];
    [self.webView.scrollView setScrollIndicatorInsets:UIEdgeInsetsMake(64, 0, 44, 0)];
    [self.webView.scrollView scrollRectToVisible:CGRectMake(0, 0, 1, 1) animated:NO];
    NSURLRequest *request = [NSURLRequest requestWithURL:self.address];
    [self.webView loadRequest:request];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - User Interactions

- (IBAction)didShareButtonPressed:(id)sender {
    if ([self.appDelegate.majorSystemVersion isEqualToString:@"8"]) {
        NSArray * data = [[NSArray alloc] init];
        data = [data arrayByAddingObject:self.address];
        data = [data arrayByAddingObject:[self.webView stringByEvaluatingJavaScriptFromString:@"document.title"]];
        UIActivityViewController *activityVC = [[UIActivityViewController alloc] initWithActivityItems:data applicationActivities:nil];
        [self presentViewController:activityVC animated:YES completion:nil];
    }
    else {
        UIActionSheet *actionSheet = [[UIActionSheet alloc] initWithTitle:nil delegate:self cancelButtonTitle:@"Cancel" destructiveButtonTitle:nil otherButtonTitles:@"Open in Safari", @"Tweet Link", @"Copy Link", @"Read It Later", nil];
        [actionSheet showInView:self.view];
    }
}

- (IBAction)didReloadOrStopButtonPressed:(UIBarButtonItem *)sender {
    //Define tag 0 as stop, tag 1 as reload
    if (sender == self.stopItem) {
        [self.webView stopLoading];
    }
    else {
        [self.webView reload];
    }
}

#pragma mark - Action Sheet Delegate

- (void)actionSheet:(UIActionSheet *)actionSheet clickedButtonAtIndex:(NSInteger)buttonIndex {
    NSString *buttonString = [actionSheet buttonTitleAtIndex:buttonIndex];
    if ([buttonString isEqualToString:@"Open in Safari"]) {
        [[UIApplication sharedApplication] openURL:self.address];
    }
    else if ([buttonString isEqualToString:@"Tweet Link"]) {
        [self performSegueWithIdentifier:@"ToNewTweet" sender:self];
    }
    else if ([buttonString isEqualToString:@"Copy Link"]) {
        UIPasteboard *sysPasterboard = [UIPasteboard generalPasteboard];
        sysPasterboard.string = self.webView.request.URL.absoluteString;
    }
    else if ([buttonString isEqualToString:@"Read It Later"]) {
        SSReadingList *readList = [SSReadingList defaultReadingList];
        [readList addReadingListItemWithURL:self.webView.request.URL title:[self.webView stringByEvaluatingJavaScriptFromString:@"document.title"] previewText:nil error:nil];
    }
}

#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"ToNewTweet"]) {
        NewTweetViewController *vc = (NewTweetViewController *)[segue.destinationViewController visibleViewController];
        vc.option = NewTweetTypeNewTweetWithText;
        vc.tweetString = self.address.description;
    }
}

#pragma mark - Web View Delegate

- (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType {
    //NSLog(@"url = %@", request.URL);
    NSURL *url = request.URL;
    if (!([url.scheme isEqualToString:@"http"] || [url.scheme isEqualToString:@"https"])) {
        if ([[UIApplication sharedApplication] canOpenURL:url]) {
            [[UIApplication sharedApplication] openURL:url];
        }
    }
    return YES;
}

- (void)webViewDidStartLoad:(UIWebView *)webView {
    self.shareButton.enabled = NO;
    NSMutableArray *itms = self.toolbar.items.mutableCopy;
    if (itms.count == 8) {
        [itms insertObject:self.stopItem atIndex:7];
    }
    else {
        itms[7] = self.stopItem;
    }
    self.toolbar.items = itms.copy;
    self.backButton.enabled = webView.canGoBack;
    self.nextButton.enabled = webView.canGoForward;
    [self.activityIndicator startAnimating];
}

- (void)webViewDidFinishLoad:(UIWebView *)webView {
    self.shareButton.enabled = YES;
    NSMutableArray *itms = self.toolbar.items.mutableCopy;
    if (itms.count == 8) {
        [itms insertObject:self.reloadItem atIndex:7];
    }
    else {
        itms[7] = self.reloadItem;
    }
    self.toolbar.items = itms.copy;
    self.title = [webView stringByEvaluatingJavaScriptFromString:@"document.title"];
    if (self.title == nil) {
        self.title = webView.request.URL.host;
    }
    [self.activityIndicator stopAnimating];
}

- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error {
    self.shareButton.enabled = NO;
    NSMutableArray *itms = self.toolbar.items.mutableCopy;
    if (itms.count == 8) {
        [itms insertObject:self.reloadItem atIndex:6];
    }
    else {
        itms[7] = self.reloadItem;
        //[itms replaceObjectAtIndex:7 withObject:self.reloadItem];
    }
    self.toolbar.items = itms.copy;
    [self.activityIndicator stopAnimating];
}

@end

File: MacThai Old Version/MacThai-iOS/MacThai-iOS/BrowserViewController.swift
//
//  BrowserViewController.swift
//  MacThai-iOS
//
//  Created by Nattapong Mos on 5/11/57.
//  Copyright (c) พ.ศ. 2557 Nattapong Mos. All rights reserved.
//

import WebKit
import UIKit

class BrowserViewController: UIViewController, UIWebViewDelegate, WKNavigationDelegate {

    @IBOutlet private var webView: UIWebView!
    var newWebView: WKWebView?
    var link: NSURLRequest?
    @IBOutlet private weak var shareButton: UIBarButtonItem!
    //@IBOutlet private weak var activityItem: UIBarButtonItem!
    //private var activityIndicator = UIActivityIndicatorView(frame: CGRectMake(0, 0, 20, 20))
    private var userDefault = NSUserDefaults()

    override internal func loadView() {
        super.loadView()
        if objc_getClass("WKWebView") != nil {
            self.newWebView = WKWebView(frame: self.webView.frame, configuration: WKWebViewConfiguration())
            self.newWebView?.navigationDelegate = self
            self.newWebView?.autoresizingMask = .FlexibleRightMargin | .FlexibleBottomMargin
            self.newWebView?.setTranslatesAutoresizingMaskIntoConstraints(false)
            //println("view subviews = \(self.view.subviews)")
            //println()
            //println("view constraint = \(self.view.constraints())")
            self.webView.removeFromSuperview()
            self.view.addSubview(self.newWebView!)
            self.view.sendSubviewToBack(self.newWebView!)
            var layout1 = NSLayoutConstraint(item: self.newWebView!, attribute: .Leading, relatedBy: .Equal, toItem: self.newWebView?.superview, attribute: .Leading, multiplier: 1, constant: 0)
            var layout2 = NSLayoutConstraint(item: self.newWebView!, attribute: .Top, relatedBy: .Equal, toItem: self.newWebView?.superview, attribute: .Top, multiplier: 1, constant: 0)
            var layout3 = NSLayoutConstraint(item: self.newWebView!.superview!, attribute: .Trailing, relatedBy: .Equal, toItem: self.newWebView, attribute: .Trailing, multiplier: 1, constant: 0)
            var layout4 = NSLayoutConstraint(item: self.newWebView!, attribute: .Bottom, relatedBy: .Equal, toItem: self.newWebView?.superview, attribute: .Bottom, multiplier: 1, constant: 0)
            self.view.addConstraints([layout1, layout2, layout3, layout4])
        }
    }

    override internal func viewDidLoad() {
        super.viewDidLoad()
        self.navigationItem.rightBarButtonItem = self.shareButton
        if let webView = self.newWebView {
            webView.scrollView.contentInset = UIEdgeInsetsMake(0, 0, 0, 0)
            webView.scrollView.scrollIndicatorInsets = UIEdgeInsetsMake(0, 0, 0, 0)
            webView.scrollView.scrollRectToVisible(CGRectMake(0, 0, 1, 1), animated: false)
        }
        else {
            self.webView.scalesPageToFit = true
            self.webView.scrollView.contentInset = UIEdgeInsetsMake(0, 0, 0, 0)
            self.webView.scrollView.scrollIndicatorInsets = UIEdgeInsetsMake(64, 0, 44, 0)
            self.webView.scrollView.scrollRectToVisible(CGRectMake(0, 0, 1, 1), animated: false)
        }
        //self.activityIndicator.activityIndicatorViewStyle = .Gray
        //self.activityItem.customView = self.activityIndicator
        if self.userDefault.objectForKey("readability") == nil {
            self.userDefault.setBool(false, forKey: "readability")
        }
        if let readabilityIsOn = self.userDefault.objectForKey("readability") as? Bool, link = self.link {
            self.title = link.URL?.absoluteString
            if readabilityIsOn {
                if let linkURL = link.URL {
                    if let url = NSURL(string: "http://www.readability.com/m?url=\(linkURL)") {
                        if let webView = self.newWebView {
                            webView.loadRequest(NSURLRequest(URL: url))
                        }
                        else {
                            self.webView.loadRequest(NSURLRequest(URL: url))
                        }
                    }
                }
            }
            else {
                if let webView = self.newWebView {
                    webView.loadRequest(link)
                }
                else {
                    self.webView.loadRequest(link)
                }
            }
        }
    }
    
    internal func webView(webView: UIWebView, shouldStartLoadWithRequest request: NSURLRequest, navigationType: UIWebViewNavigationType) -> Bool {
        if request.URL?.host == "www.macthai.com" {
            var pathCompo = request.URL?.pathComponents
            if pathCompo?[1] as? String == "category" {
                if let urlString = request.URL?.absoluteString?.stringByAppendingPathComponent("feed") {
                    var sendData = NSURL(string: urlString)
                    performSegueWithIdentifier("ToSubFeed", sender: sendData)
                }
                return false
            }
            else if pathCompo?[1] as? String == "tag" {
                if let urlString = request.URL?.absoluteString?.stringByAppendingPathComponent("feed") {
                    var sendData = NSURL(string: urlString)
                    performSegueWithIdentifier("ToSubFeed", sender: sendData)
                }
                return false
            }
            else if pathCompo?[1] as? String == "author" {
                if let urlString = request.URL?.absoluteString?.stringByAppendingPathComponent("feed") {
                    var sendData = NSURL(string: urlString)
                    performSegueWithIdentifier("ToSubFeed", sender: sendData)
                }
                return false
            }
        }
        return true
    }

    // MARK: - Web View Delegate

    internal func webViewDidStartLoad(webView: UIWebView) {
        //self.activityIndicator.startAnimating()
        UIApplication.sharedApplication().networkActivityIndicatorVisible = true
    }

    internal func webViewDidFinishLoad(webView: UIWebView) {
        NSTimer.scheduledTimerWithTimeInterval(2, target: self, selector: "enableButton", userInfo: nil, repeats: false)
        //self.activityIndicator.stopAnimating()
        UIApplication.sharedApplication().networkActivityIndicatorVisible = false
        self.title = webView.stringByEvaluatingJavaScriptFromString("document.title")
    }
    
    internal func webView(webView: UIWebView, didFailLoadWithError error: NSError) {
        UIApplication.sharedApplication().networkActivityIndicatorVisible = false
        //self.activityIndicator.stopAnimating()
    }
    
    // MARK: - User Interactions
    
    @IBAction internal func didShareButtonPressed (sender: UIBarButtonItem!) {
        var data = [AnyObject]()
        if let webView = self.newWebView {
            if let url = self.newWebView?.URL {
                data.append(url)
            }
            webView.evaluateJavaScript("document.title", completionHandler: { (result, error) -> Void in
                if error == nil {
                    if var resultString = result as? String {
                        resultString = resultString.stringByReplacingOccurrencesOfString(" | Macthai.com", withString: " @macthainews")
                        data.append(resultString)
                        var activity = UIActivityViewController(activityItems: data, applicationActivities: [OpenInSafariActivity()])
                        self.presentViewController(activity, animated: true, completion: nil)
                    }
                }
            })
        }
        else {
            if let url = self.webView.stringByEvaluatingJavaScriptFromString("window.location") {
                data.append(url)
            }
            if let title = self.webView.stringByEvaluatingJavaScriptFromString("document.title")?.stringByReplacingOccurrencesOfString(" | Macthai.com", withString: " @macthainews") {
                data.append(title)
            }
            var activity = UIActivityViewController(activityItems: data, applicationActivities: [OpenInSafariActivity()])
            presentViewController(activity, animated: true, completion: nil)
        }
    }

    override internal func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Navigation

    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        let segueID = segue.identifier
        if segueID == "ToSubFeed" {
            let vc = segue.destinationViewController as! SubFeedViewController
            vc.url = sender as? NSURL
        }
        else if segueID == "ToBrowser" {
            let vc = segue.destinationViewController as! BrowserViewController
            if let sender = sender as? NSURLRequest {
                vc.link = sender
            }
        }
    }

    // MARK: - WebKit Navigation Delegate
    
    func webView(webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        UIApplication.sharedApplication().networkActivityIndicatorVisible = true
    }

    func webView(webView: WKWebView, didFinishNavigation navigation: WKNavigation!) {
        UIApplication.sharedApplication().networkActivityIndicatorVisible = false
        self.title = webView.title
    }

    func webView(webView: WKWebView, decidePolicyForNavigationResponse navigationAction: WKNavigationResponse, decisionHandler: (WKNavigationResponsePolicy) -> Void) {
        if webView.URL?.host == "www.macthai.com" {
            var pathCompo = webView.URL?.pathComponents
            if pathCompo?[1] as? String == "category" {
                decisionHandler(.Cancel)
                if let urlString = webView.URL?.absoluteString?.stringByAppendingPathComponent("feed") {
                    var sendData = NSURL(string: urlString)
                    performSegueWithIdentifier("ToSubFeed", sender: sendData)
                }
            }
            else if pathCompo?[1] as? String == "tag" {
                decisionHandler(.Cancel)
                if let urlString = webView.URL?.absoluteString?.stringByAppendingPathComponent("feed") {
                    var sendData = NSURL(string: urlString)
                    performSegueWithIdentifier("ToSubFeed", sender: sendData)
                }
            }
            else if pathCompo?[1] as? String == "author" {
                decisionHandler(.Cancel)
                if let urlString = webView.URL?.absoluteString?.stringByAppendingPathComponent("feed") {
                    var sendData = NSURL(string: urlString)
                    performSegueWithIdentifier("ToSubFeed", sender: sendData)
                }
            }
        }
        decisionHandler(.Allow)
    }

}
